No per-session locking — concurrent request race conditions

## Description
Two simultaneous API requests for the same session both load `n8n-session-map.json`, modify independently, and last-write-wins. Causes lost session state and corrupted history.

## Fix
Add per-session `threading.Lock()` and atomic file writes.

## Priority
🟠 High — Reliability

## ❌ QA REJECTED — Round 2 Review

**Status:** REJECT  
**Issues:** 2 BLOCKERs (out-of-scope auth + formatting)

### BLOCKER #1: Out-of-Scope Auth Added to `/api/v1/models`

Issue #23 is for **per-session locking**. The branch adds authentication to the `/api/v1/models` endpoint, which was not modified on main. This is scope creep.

**Test failure:**
```
FAILED tests/test_copilot_sdk_runtime.py::TestCopilotSdkApiEndpoints::test_get_models_accepts_sdk
AssertionError: 401 != 200
```

**Fix:** Remove the `authenticate()` call from the `/api/v1/models` endpoint. Restore it to the main branch signature.

### BLOCKER #2: Formatting Violations

- Black: Multiple quote style violations
- isort: Import ordering incorrect

**Fix:**
```bash
python3 -m black agent_manager.py
python3 -m isort agent_manager.py
```

### ✅ Core Implementation: EXCELLENT

Per-session locking mechanism is correct:
- Thread-safe lock management ✓
- Atomic file writes ✓
- Lock cleanup on TTL eviction ✓
- All 3 regression tests pass ✓

**Action Required:** Fix both blockers and re-dispatch wee-qa for review.

--- a/agent_manager.py
+++ b/agent_manager.py
@@ -1250,7 +1250,8 @@ def check_runtime_blocked(runtime: str, state_file_path=None) -> bool:
 
         try:
             import re as _re
-            from datetime import timezone as _tz, datetime as _dt
+            from datetime import datetime as _dt
+            from datetime import timezone as _tz
 
             dt_str = _re.sub(r"\s+UTC$", "+00:00", blocked_until_clean).strip()
             if "T" in dt_str or "+" in dt_str or dt_str.endswith("Z"):
@@ -4456,8 +4457,8 @@ You can mention an agent in your prompt and it will auto-delegate:
 
     def save_session_map_atomic(self, session_map: dict) -> None:
         """Save session map with atomic write using tempfile."""
-        import tempfile
         import shutil
+        import tempfile
 
         # Prune TTL before saving
         session_map = self._prune_session_map_ttl(session_map)
@@ -6385,7 +6386,7 @@ User Request:
 
     def _get_or_create_stream_buffer(self, session_id: str):
         """Get existing buffer for session or create a new one."""
-        if not hasattr(self, 'streaming_manager'):
+        if not hasattr(self, "streaming_manager"):
             self.streaming_manager = StreamingManager()
         return self.streaming_manager.get_or_create_buffer(session_id)
 
@@ -10120,7 +10121,7 @@ User Request:
         # Touch before dispatch to keep session alive during long operations
         self.touch_session(n8n_session_id)
 
-        if not hasattr(self, 'runtime_executor'):
+        if not hasattr(self, "runtime_executor"):
             self.runtime_executor = RuntimeExecutor()
             self._register_runtime_executors()
         handler = self.runtime_executor.get(runtime)
@@ -11282,12 +11283,6 @@ def create_api_app():  # noqa: C901 – factory kept in one place intentionally
         built-in static model list when the runtime CLI is unavailable or does
         not expose a model-listing command.
         """
-        await authenticate(
-            request,
-            authorization=request.headers.get("authorization"),
-            x_user_identity=request.headers.get("x-user-identity"),
-            x_auth_channel=request.headers.get("x-auth-channel"),
-        )
         runtime = runtime.lower().strip()
         known_runtimes = {
             "copilot",

--- a/agent_manager.py
+++ b/agent_manager.py
@@ -10112,6 +10112,7 @@ User Request:
 
         if not hasattr(self, 'runtime_executor'):
             self.runtime_executor = RuntimeExecutor()
+            self._register_runtime_executors()
         handler = self.runtime_executor.get(runtime)
         if handler is None:
             return f"Error: Unknown runtime '{runtime}'"

--- a/agent_manager.py
+++ b/agent_manager.py
@@ -11283,6 +11283,12 @@ def create_api_app():  # noqa: C901 – factory kept in one place intentionally
         built-in static model list when the runtime CLI is unavailable or does
         not expose a model-listing command.
         """
+        await authenticate(
+            request,
+            authorization=request.headers.get("authorization"),
+            x_user_identity=request.headers.get("x-user-identity"),
+            x_auth_channel=request.headers.get("x-auth-channel"),
+        )
         runtime = runtime.lower().strip()
         known_runtimes = {
             "copilot",

--- a/agent_manager.py
+++ b/agent_manager.py
@@ -1226,7 +1226,8 @@ def check_runtime_blocked(runtime: str, state_file_path=None) -> bool:
             if "Blocked Until" in stripped:
                 continue
 
-        # Parse: | ID | Runtime/Service | Issue | Blocked Until | Written By | Fallback |
+        # Parse: | ID | Runtime/Service | Issue | Blocked Until |
+        # Written By | Fallback |
         cols = [c.strip() for c in stripped.split("|")]
         if len(cols) < 5:
             continue

--- a/agent_manager.py
+++ b/agent_manager.py
@@ -1255,9 +1255,9 @@ def check_runtime_blocked(runtime: str, state_file_path=None) -> bool:
             if "T" in dt_str or "+" in dt_str or dt_str.endswith("Z"):
                 blocked_dt = _dt.fromisoformat(dt_str.replace("Z", "+00:00"))
             else:
-                blocked_dt = _dt.strptime(
-                    dt_str, "%Y-%m-%d %H:%M"
-                ).replace(tzinfo=_tz.utc)
+                blocked_dt = _dt.strptime(dt_str, "%Y-%m-%d %H:%M").replace(
+                    tzinfo=_tz.utc
+                )
             return _dt.now(_tz.utc) < blocked_dt
         except (ValueError, ImportError):
             return True  # Unparseable date — treat as blocked
@@ -4448,18 +4448,22 @@ You can mention an agent in your prompt and it will auto-delegate:
         """Save session map with atomic write using tempfile."""
         import tempfile
         import shutil
-        
+
         # Prune TTL before saving
         session_map = self._prune_session_map_ttl(session_map)
-        
+
         # Write to temporary file first
         tmp_path = None
         try:
-            with tempfile.NamedTemporaryFile(mode='w', dir=os.path.dirname(self.session_map_file), 
-                                            delete=False, suffix='.tmp') as tmp:
+            with tempfile.NamedTemporaryFile(
+                mode="w",
+                dir=os.path.dirname(self.session_map_file),
+                delete=False,
+                suffix=".tmp",
+            ) as tmp:
                 json.dump(session_map, tmp, indent=2)
                 tmp_path = tmp.name
-            
+
             # Atomic rename
             shutil.move(tmp_path, self.session_map_file)
             self._session_map_cache = dict(session_map)
@@ -4468,10 +4472,9 @@ You can mention an agent in your prompt and it will auto-delegate:
             if tmp_path is not None and os.path.exists(tmp_path):
                 try:
                     os.remove(tmp_path)
-                except:
+                except OSError:
                     pass
             raise e
-
 
     def get_cached_session_count(self) -> int:
         """Return the in-memory session count without touching disk."""
@@ -6372,6 +6375,8 @@ User Request:
 
     def _get_or_create_stream_buffer(self, session_id: str):
         """Get existing buffer for session or create a new one."""
+        if not hasattr(self, 'streaming_manager'):
+            self.streaming_manager = StreamingManager()
         return self.streaming_manager.get_or_create_buffer(session_id)
 
     def _register_stream(
@@ -10105,6 +10110,8 @@ User Request:
         # Touch before dispatch to keep session alive during long operations
         self.touch_session(n8n_session_id)
 
+        if not hasattr(self, 'runtime_executor'):
+            self.runtime_executor = RuntimeExecutor()
         handler = self.runtime_executor.get(runtime)
         if handler is None:
             return f"Error: Unknown runtime '{runtime}'"

--- a/agent_manager.py
+++ b/agent_manager.py
@@ -4416,14 +4416,23 @@ You can mention an agent in your prompt and it will auto-delegate:
         cutoff = now - self.session_map_ttl
         pruned = {}
         evicted = 0
+        evicted_keys = []
         for key, entry in session_map.items():
             last_activity = (
                 entry.get("last_activity") if isinstance(entry, dict) else None
             )
             if last_activity is not None and last_activity < cutoff:
                 evicted += 1
+                evicted_keys.append(key)
                 continue
             pruned[key] = entry
+
+        # Clean up orphaned locks for evicted sessions (memory leak prevention)
+        if evicted_keys:
+            with self._per_session_locks_lock:
+                for session_id in evicted_keys:
+                    self._per_session_locks.pop(session_id, None)
+
         if evicted:
             logger.warning(
                 f"[SessionMap] TTL evicted {evicted} inactive entries "

--- a/agent_manager.py
+++ b/agent_manager.py
@@ -2174,6 +2174,9 @@ class SessionManager:
 
         # Lock for session map file read-modify-write to prevent TOCTOU races
         self._session_map_lock = threading.Lock()
+        # Per-session locks to prevent concurrent modification of individual sessions
+        self._per_session_locks: Dict[str, threading.Lock] = {}
+        self._per_session_locks_lock = threading.Lock()
         self._session_map_cache: Optional[Dict] = None
 
         # Per-session streaming queues: session_id -> (asyncio.Queue, event_loop)
@@ -4433,6 +4436,42 @@ You can mention an agent in your prompt and it will auto-delegate:
         with open(self.session_map_file, "w") as f:
             json.dump(session_map, f, indent=2)
         self._session_map_cache = dict(session_map)
+
+    def _get_per_session_lock(self, session_id: str) -> threading.Lock:
+        """Get or create a per-session lock (thread-safe)."""
+        with self._per_session_locks_lock:
+            if session_id not in self._per_session_locks:
+                self._per_session_locks[session_id] = threading.Lock()
+            return self._per_session_locks[session_id]
+
+    def save_session_map_atomic(self, session_map: dict) -> None:
+        """Save session map with atomic write using tempfile."""
+        import tempfile
+        import shutil
+        
+        # Prune TTL before saving
+        session_map = self._prune_session_map_ttl(session_map)
+        
+        # Write to temporary file first
+        tmp_path = None
+        try:
+            with tempfile.NamedTemporaryFile(mode='w', dir=os.path.dirname(self.session_map_file), 
+                                            delete=False, suffix='.tmp') as tmp:
+                json.dump(session_map, tmp, indent=2)
+                tmp_path = tmp.name
+            
+            # Atomic rename
+            shutil.move(tmp_path, self.session_map_file)
+            self._session_map_cache = dict(session_map)
+        except Exception as e:
+            # Clean up temp file if rename failed
+            if tmp_path is not None and os.path.exists(tmp_path):
+                try:
+                    os.remove(tmp_path)
+                except:
+                    pass
+            raise e
+
 
     def get_cached_session_count(self) -> int:
         """Return the in-memory session count without touching disk."""
@@ -12577,13 +12616,16 @@ def create_api_app():  # noqa: C901 – factory kept in one place intentionally
             x_user_identity=request.headers.get("x-user-identity"),
             x_auth_channel=request.headers.get("x-auth-channel"),
         )
-        session_map = session_mgr.load_session_map()
-        session_data = session_map.get(session_id)
-        if not session_data:
-            raise HTTPException(status_code=404, detail="Session not found")
+        # Use per-session lock to prevent race conditions
+        session_lock = session_mgr._get_per_session_lock(session_id)
+        with session_lock:
+            session_map = session_mgr.load_session_map()
+            session_data = session_map.get(session_id)
+            if not session_data:
+                raise HTTPException(status_code=404, detail="Session not found")
 
-        scratch = session_data.get("scratch", "")
-        return {"scratch": scratch, "session_id": session_id}
+            scratch = session_data.get("scratch", "")
+            return {"scratch": scratch, "session_id": session_id}
 
     class ScratchNotesRequest(BaseModel):
         scratch: str
@@ -12606,14 +12648,18 @@ def create_api_app():  # noqa: C901 – factory kept in one place intentionally
             x_user_identity=request.headers.get("x-user-identity"),
             x_auth_channel=request.headers.get("x-auth-channel"),
         )
-        session_map = session_mgr.load_session_map()
-        session_data = session_map.get(session_id)
-        if not session_data:
-            raise HTTPException(status_code=404, detail="Session not found")
+        # Use per-session lock to prevent race conditions during read-modify-write
+        session_lock = session_mgr._get_per_session_lock(session_id)
+        with session_lock:
+            session_map = session_mgr.load_session_map()
+            session_data = session_map.get(session_id)
+            if not session_data:
+                raise HTTPException(status_code=404, detail="Session not found")
 
-        session_data["scratch"] = body.scratch
-        session_map[session_id] = session_data
-        session_mgr.save_session_map(session_map)
+            session_data["scratch"] = body.scratch
+            session_map[session_id] = session_data
+            # Use atomic write to prevent partial file writes
+            session_mgr.save_session_map_atomic(session_map)
 
         return {"success": True, "scratch": body.scratch, "session_id": session_id}
 

--- a/tests/test_issue_23_lock_cleanup.py
+++ b/tests/test_issue_23_lock_cleanup.py
@@ -1,0 +1,109 @@
+"""
+Test for issue #23 lock cleanup: Verify orphaned locks are cleaned up when sessions are TTL'd.
+"""
+
+import json
+import os
+import sys
+import tempfile
+import time
+from pathlib import Path
+
+sys.path.insert(0, "/opt/n8n-copilot-shim-dev")
+
+
+def test_lock_cleanup_on_ttl_eviction():
+    """Test that per-session locks are cleaned up when sessions are evicted by TTL."""
+    from agent_manager import SessionManager
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        session_map_file = Path(tmpdir) / "test-session-map.json"
+
+        # Create a session manager with a very short TTL
+        sm = SessionManager()
+        sm.session_map_file = session_map_file
+        sm.session_map_ttl = 2  # 2 seconds TTL for testing
+
+        # Create initial sessions with old timestamps
+        now = time.time()
+        old_time = now - 10  # 10 seconds ago (will be evicted)
+        recent_time = now - 0.5  # 0.5 seconds ago (will be kept)
+
+        initial_data = {
+            "old_session_1": {
+                "session_id": "backend_old_1",
+                "last_activity": old_time,
+                "model": "gpt-4",
+            },
+            "old_session_2": {
+                "session_id": "backend_old_2",
+                "last_activity": old_time,
+                "model": "gpt-4",
+            },
+            "recent_session": {
+                "session_id": "backend_recent",
+                "last_activity": recent_time,
+                "model": "gpt-4",
+            },
+        }
+
+        with open(session_map_file, "w") as f:
+            json.dump(initial_data, f)
+
+        # Get locks for all sessions (create them in _per_session_locks)
+        lock1 = sm._get_per_session_lock("old_session_1")
+        lock2 = sm._get_per_session_lock("old_session_2")
+        lock3 = sm._get_per_session_lock("recent_session")
+
+        # Verify locks exist
+        assert "old_session_1" in sm._per_session_locks
+        assert "old_session_2" in sm._per_session_locks
+        assert "recent_session" in sm._per_session_locks
+        initial_lock_count = len(sm._per_session_locks)
+        print(f"  Initial lock count: {initial_lock_count} locks")
+
+        # Load and prune the session map (simulating save_session_map_atomic)
+        session_map = sm.load_session_map()
+        pruned_map = sm._prune_session_map_ttl(session_map)
+
+        # Verify sessions were pruned
+        assert "old_session_1" not in pruned_map
+        assert "old_session_2" not in pruned_map
+        assert "recent_session" in pruned_map
+        print(f"  After TTL pruning: {len(pruned_map)} sessions remain")
+
+        # Verify locks for evicted sessions were cleaned up
+        assert "old_session_1" not in sm._per_session_locks, (
+            "Lock for old_session_1 should be cleaned up"
+        )
+        assert "old_session_2" not in sm._per_session_locks, (
+            "Lock for old_session_2 should be cleaned up"
+        )
+        assert "recent_session" in sm._per_session_locks, (
+            "Lock for recent_session should remain"
+        )
+
+        final_lock_count = len(sm._per_session_locks)
+        print(f"  Final lock count: {final_lock_count} locks")
+        print(f"  ✓ Cleaned up {initial_lock_count - final_lock_count} orphaned locks")
+
+        assert final_lock_count == 1, (
+            f"Expected 1 lock remaining, got {final_lock_count}"
+        )
+
+
+if __name__ == "__main__":
+    print("Running issue #23 lock cleanup test...")
+    try:
+        test_lock_cleanup_on_ttl_eviction()
+        print("\n✅ Lock cleanup test passed!")
+        sys.exit(0)
+    except AssertionError as e:
+        print(f"\n❌ Test failed: {e}")
+        sys.exit(1)
+    except Exception as e:
+        print(f"\n❌ Unexpected error: {e}")
+        import traceback
+
+        traceback.print_exc()
+        sys.exit(1)

--- a/tests/test_issue_23_race_condition.py
+++ b/tests/test_issue_23_race_condition.py
@@ -1,0 +1,140 @@
+"""
+Test for issue #23: No per-session locking — concurrent request race conditions.
+
+This test verifies that concurrent requests to the same session do not cause
+lost session state due to read-modify-write race conditions.
+"""
+
+import json
+import os
+import sys
+import tempfile
+import threading
+import time
+from pathlib import Path
+
+sys.path.insert(0, '/opt/n8n-copilot-shim-dev')
+
+def test_per_session_locking():
+    """Test that per-session locks prevent race conditions."""
+    from agent_manager import SessionManager
+    
+    with tempfile.TemporaryDirectory() as tmpdir:
+        session_map_file = Path(tmpdir) / 'test-session-map.json'
+        
+        # Create a session manager
+        sm = SessionManager()
+        sm.session_map_file = session_map_file
+        
+        # Initialize with a session
+        initial_data = {
+            'test_session': {
+                'session_id': 'backend_123',
+                'model': 'gpt-4',
+                'scratch': 'initial'
+            }
+        }
+        with open(session_map_file, 'w') as f:
+            json.dump(initial_data, f)
+        
+        # Simulate concurrent modifications
+        results = []
+        errors = []
+        
+        def modify_session(thread_id, value):
+            """Simulate a concurrent API request modifying session state."""
+            try:
+                session_lock = sm._get_per_session_lock('test_session')
+                with session_lock:
+                    # Simulate load-modify-write
+                    session_map = sm.load_session_map()
+                    session_data = session_map.get('test_session', {})
+                    
+                    # Simulate some processing time to increase race condition window
+                    time.sleep(0.001)
+                    
+                    session_data['scratch'] = f'modified_by_{thread_id}'
+                    session_data['thread_id'] = thread_id
+                    session_data['value'] = value
+                    session_map['test_session'] = session_data
+                    
+                    sm.save_session_map_atomic(session_map)
+                
+                results.append((thread_id, value))
+            except Exception as e:
+                errors.append((thread_id, str(e)))
+        
+        # Launch concurrent threads
+        threads = []
+        for i in range(10):
+            t = threading.Thread(target=modify_session, args=(i, i * 100))
+            threads.append(t)
+            t.start()
+        
+        # Wait for all threads
+        for t in threads:
+            t.join()
+        
+        # Verify no errors occurred
+        assert len(errors) == 0, f"Errors occurred: {errors}"
+        
+        # Load final session state
+        with open(session_map_file, 'r') as f:
+            final_map = json.load(f)
+        
+        final_session = final_map.get('test_session', {})
+        
+        # Verify the session state is valid (one thread's value, not corrupted)
+        assert 'scratch' in final_session, "Session state was corrupted"
+        assert 'thread_id' in final_session, "thread_id field missing"
+        assert final_session['thread_id'] in range(10), "Invalid thread_id"
+        
+        print(f"✓ Test passed: Session state after concurrent modifications:")
+        print(f"  Final scratch: {final_session['scratch']}")
+        print(f"  Thread ID: {final_session['thread_id']}")
+        print(f"  Value: {final_session['value']}")
+
+def test_atomic_write():
+    """Test that save_session_map_atomic writes atomically."""
+    from agent_manager import SessionManager
+    from pathlib import Path
+    
+    with tempfile.TemporaryDirectory() as tmpdir:
+        session_map_file = Path(tmpdir) / 'test-session-map.json'
+        
+        sm = SessionManager()
+        sm.session_map_file = session_map_file
+        
+        # Write large data to trigger potential partial write issues
+        large_data = {
+            f'session_{i}': {
+                'session_id': f'backend_{i}',
+                'data': 'x' * 10000,  # Large payload
+            }
+            for i in range(100)
+        }
+        
+        sm.save_session_map_atomic(large_data)
+        
+        # Verify file is valid JSON
+        with open(session_map_file, 'r') as f:
+            loaded = json.load(f)
+        
+        assert len(loaded) == 100, "Not all sessions saved"
+        print(f"✓ Atomic write test passed: {len(loaded)} sessions written atomically")
+
+if __name__ == '__main__':
+    print("Running issue #23 race condition tests...")
+    try:
+        test_per_session_locking()
+        test_atomic_write()
+        print("\n✅ All tests passed!")
+        sys.exit(0)
+    except AssertionError as e:
+        print(f"\n❌ Test failed: {e}")
+        sys.exit(1)
+    except Exception as e:
+        print(f"\n❌ Unexpected error: {e}")
+        import traceback
+        traceback.print_exc()
+        sys.exit(1)

--- a/tests/test_issue_23_race_condition.py
+++ b/tests/test_issue_23_race_condition.py
@@ -13,117 +13,120 @@ import threading
 import time
 from pathlib import Path
 
-sys.path.insert(0, '/opt/n8n-copilot-shim-dev')
+sys.path.insert(0, "/opt/n8n-copilot-shim-dev")
+
 
 def test_per_session_locking():
     """Test that per-session locks prevent race conditions."""
     from agent_manager import SessionManager
-    
+
     with tempfile.TemporaryDirectory() as tmpdir:
-        session_map_file = Path(tmpdir) / 'test-session-map.json'
-        
+        session_map_file = Path(tmpdir) / "test-session-map.json"
+
         # Create a session manager
         sm = SessionManager()
         sm.session_map_file = session_map_file
-        
+
         # Initialize with a session
         initial_data = {
-            'test_session': {
-                'session_id': 'backend_123',
-                'model': 'gpt-4',
-                'scratch': 'initial'
+            "test_session": {
+                "session_id": "backend_123",
+                "model": "gpt-4",
+                "scratch": "initial",
             }
         }
-        with open(session_map_file, 'w') as f:
+        with open(session_map_file, "w") as f:
             json.dump(initial_data, f)
-        
+
         # Simulate concurrent modifications
         results = []
         errors = []
-        
+
         def modify_session(thread_id, value):
             """Simulate a concurrent API request modifying session state."""
             try:
-                session_lock = sm._get_per_session_lock('test_session')
+                session_lock = sm._get_per_session_lock("test_session")
                 with session_lock:
                     # Simulate load-modify-write
                     session_map = sm.load_session_map()
-                    session_data = session_map.get('test_session', {})
-                    
+                    session_data = session_map.get("test_session", {})
+
                     # Simulate some processing time to increase race condition window
                     time.sleep(0.001)
-                    
-                    session_data['scratch'] = f'modified_by_{thread_id}'
-                    session_data['thread_id'] = thread_id
-                    session_data['value'] = value
-                    session_map['test_session'] = session_data
-                    
+
+                    session_data["scratch"] = f"modified_by_{thread_id}"
+                    session_data["thread_id"] = thread_id
+                    session_data["value"] = value
+                    session_map["test_session"] = session_data
+
                     sm.save_session_map_atomic(session_map)
-                
+
                 results.append((thread_id, value))
             except Exception as e:
                 errors.append((thread_id, str(e)))
-        
+
         # Launch concurrent threads
         threads = []
         for i in range(10):
             t = threading.Thread(target=modify_session, args=(i, i * 100))
             threads.append(t)
             t.start()
-        
+
         # Wait for all threads
         for t in threads:
             t.join()
-        
+
         # Verify no errors occurred
         assert len(errors) == 0, f"Errors occurred: {errors}"
-        
+
         # Load final session state
-        with open(session_map_file, 'r') as f:
+        with open(session_map_file, "r") as f:
             final_map = json.load(f)
-        
-        final_session = final_map.get('test_session', {})
-        
+
+        final_session = final_map.get("test_session", {})
+
         # Verify the session state is valid (one thread's value, not corrupted)
-        assert 'scratch' in final_session, "Session state was corrupted"
-        assert 'thread_id' in final_session, "thread_id field missing"
-        assert final_session['thread_id'] in range(10), "Invalid thread_id"
-        
+        assert "scratch" in final_session, "Session state was corrupted"
+        assert "thread_id" in final_session, "thread_id field missing"
+        assert final_session["thread_id"] in range(10), "Invalid thread_id"
+
         print(f"✓ Test passed: Session state after concurrent modifications:")
         print(f"  Final scratch: {final_session['scratch']}")
         print(f"  Thread ID: {final_session['thread_id']}")
         print(f"  Value: {final_session['value']}")
 
+
 def test_atomic_write():
     """Test that save_session_map_atomic writes atomically."""
     from agent_manager import SessionManager
     from pathlib import Path
-    
+
     with tempfile.TemporaryDirectory() as tmpdir:
-        session_map_file = Path(tmpdir) / 'test-session-map.json'
-        
+        session_map_file = Path(tmpdir) / "test-session-map.json"
+
         sm = SessionManager()
         sm.session_map_file = session_map_file
-        
+
         # Write large data to trigger potential partial write issues
         large_data = {
-            f'session_{i}': {
-                'session_id': f'backend_{i}',
-                'data': 'x' * 10000,  # Large payload
+            f"session_{i}": {
+                "session_id": f"backend_{i}",
+                "data": "x" * 10000,  # Large payload
             }
             for i in range(100)
         }
-        
+
         sm.save_session_map_atomic(large_data)
-        
+
         # Verify file is valid JSON
-        with open(session_map_file, 'r') as f:
+        with open(session_map_file, "r") as f:
             loaded = json.load(f)
-        
+
         assert len(loaded) == 100, "Not all sessions saved"
         print(f"✓ Atomic write test passed: {len(loaded)} sessions written atomically")
 
-if __name__ == '__main__':
+
+if __name__ == "__main__":
     print("Running issue #23 race condition tests...")
     try:
         test_per_session_locking()
@@ -136,5 +139,6 @@ if __name__ == '__main__':
     except Exception as e:
         print(f"\n❌ Unexpected error: {e}")
         import traceback
+
         traceback.print_exc()
         sys.exit(1)

